### PR TITLE
Codemod: Migrate to test package

### LINKED ISSUE #25429: Automigration from @storybook/<jest|testing-library> to @storybook/test

### PR DESCRIPTION
**Pull request by kasperpeulen**
Closes #25429

## What I did

See issue

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-25958-sha-570b3ba2`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-25958-sha-570b3ba2 sandbox` or in an existing project with `npx storybook@0.0.0-pr-25958-sha-570b3ba2 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-25958-sha-570b3ba2`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-25958-sha-570b3ba2) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`kasper/test-codemod`](https://github.com/storybookjs/storybook/tree/kasper/test-codemod) |
| **Commit** | [`570b3ba2`](https://github.com/storybookjs/storybook/commit/570b3ba2ab2fa763b8f3159bc13b5524f0df8bfa) |
| **Datetime** | Fri Feb  9 11:34:13 UTC 2024 (`1707478453`) |
| **Workflow run** | [7843367070](https://github.com/storybookjs/storybook/actions/runs/7843367070) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=25958`_
</details>

### CODE REVIEW

File: code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.test.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+}) => {
+  return removeJestTestingLibrary.check({
+    packageManager: packageManager as any,
+    configDir: '',
+    mainConfig: mainConfig as any,
+    storybookVersion,
+  });
+};
+
+it('should show incompatible addons registered in main.js', async () => {
```
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.test.ts:24]: Is it really about the addons which are registerd in main.js. It's rather about installed dependencies, isn't it. So the title of this test doesn't quite fit.

File: code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  expect(await tsTransform(input)).toMatchInlineSnapshot(`
+    import { expect } from '@storybook/test';
+    import * as test from '@storybook/test';
+    import { within, userEvent } from '@storybook/test';
+
+    const onFocusMock = test.fn();
+    const onSearchMock = test.fn();
+
+    test.spyOn(window, 'Something');
```
Comment (valentinpalkovic) [inline code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts:44]: If I run the tests locally, it only matches if the snapshot looks like this:

```tsx
import * as test, { expect } from "@storybook/test";
import * as test from "@storybook/test";
import { within, userEvent } from "@storybook/test";

const onFocusMock = test.fn();
const onSearchMock = test.fn();

test.spyOn(window, 'Something');
```

File: code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        'Attention'
+      )}: We've detected that you're using the following packages which are known to be incompatible with Storybook 8:
+
+      ${incompatiblePackages.map((name) => `- ${chalk.cyan(`${name}`)}`).join('\n')}
+      
+      Install the replacement for those packages: ${chalk.cyan('@storybook/test')}
+      
+      And run the following codemod:
+       ${chalk.cyan(
+         'npx sb migrate migrate-to-test-package --glob="**/*.stories.@(js|jsx|ts|tsx)"'
```
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.ts:28]: As far as I know we shouldn't recommend `sb`, but `storybook` instead.

```suggestion
         'npx storybook@latest migrate migrate-to-test-package --glob="**/*.stories.@(js|jsx|ts|tsx)"'
```

File: code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+test('replace jest and testing-library with the test package', async () => {
+  const input = dedent`
+    import { expect } from '@storybook/jest';
+    import { within, userEvent } from '@storybook/testing-library';
+  `;
+
+  expect(await tsTransform(input)).toMatchInlineSnapshot(`
+    import { expect } from '@storybook/test';
+    import { within, userEvent } from '@storybook/test';
```
Comment (valentinpalkovic) [inline code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts:21]: I guess merging both lines is possible but might require some further action?

File: code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts (hunk 2 above)
Comment (valentinpalkovic) [reply code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts:44]: I can verify this behaviour running the codemod on https://github.com/washingtonpost/wpds-ui-kit

image

File: code/lib/codemod/src/transforms/migrate-to-test-package.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  });
+
+  let output = printCsf(csf).code;
+  try {
+    output = await prettier.format(output, {
+      ...(await prettier.resolveConfig(info.path)),
+      filepath: info.path,
+    });
+  } catch (e) {
+    console.warn(`Failed applying prettier to ${info.path}.`);
```
Comment (valentinpalkovic) [inline code/lib/codemod/src/transforms/migrate-to-test-package.ts:59]: Prettier sometimes fails. Example content on which it failed:
https://github.com/washingtonpost/wpds-ui-kit/blob/main/ui/input-text/src/play.stories.tsx

File: code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.test.ts (hunk 1 above)
Comment (kasperpeulen) [reply code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.test.ts]: fixed

File: code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts (hunk 2 above)
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts:44]: Fixed

File: code/lib/codemod/src/transforms/migrate-to-test-package.ts (hunk 5 above)
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/migrate-to-test-package.ts:58]: Fixed

File: code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.ts (hunk 3 above)
Comment (kasperpeulen) [reply code/lib/cli/src/automigrate/fixes/remove-jest-testing-library.ts]: Discussed with @shilman, but in this case it is better to use the local `storybook` binary, as if we would delete this codemod in feature version, it won't be there. But the local codemod will always be there.

File: code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts (hunk 4 above)
Comment (kasperpeulen) [reply code/lib/codemod/src/transforms/__tests__/migrate-to-test-package.test.ts:21]: Yeah, I think people that care about this, will have a eslint rule for merging this automatically.

**Review by valentinpalkovic — APPROVED** [review]

**Pull request merged**